lib/AST/Module.cpp

```diff
@@ -4133,3 +4098,46 @@ const UnifiedStatsReporter::TraceFormatter*
 FrontendStatsTracer::getTraceFormatter<const SourceFile *>() {
   return &TF;
 }
+
+static bool prefixMatches(StringRef prefix, StringRef path) {
+  auto prefixIt = llvm::sys::path::begin(prefix),
+       prefixEnd = llvm::sys::path::end(prefix);
+  for (auto pathIt = llvm::sys::path::begin(path),
+            pathEnd = llvm::sys::path::end(path);
+       prefixIt != prefixEnd && pathIt != pathEnd; ++prefixIt, ++pathIt) {
+    if (*prefixIt != *pathIt)
+      return false;
+  }
+  return prefixIt == prefixEnd;
+}
+
+bool IsNonUserModuleRequest::evaluate(Evaluator &evaluator, ModuleDecl *mod) const {
+  // For clang submodules, retrieve their top level module (submodules won't
+  // have any source filename)
+  mod = mod->getTopLevelModule();
```

Comment:
Would it be worth recursing into `isNonUserModule()` on the top-level module so that it only has to be computed once for all the submodules that share a common top-level module?

Reply:
Heh, I did do that and changed my mind. Sure, will change back.